lnwallet/musig_session.go

```diff
@@ -0,0 +1,535 @@
+	// generate it now ourselves.
+	m.combinedNonce, err = musig2.AggregateNonces([][musig2.PubNonceSize]byte{
+		m.nonces.SigningNonce.PubNonce,
+		m.nonces.VerificationNonce.PubNonce,
+	})
+	if err != nil {
+		return nil
+	}
+
+	return nil
+}
+
+// taprootKeyspendSighash generates the sighash for a taproot key spend. As
+// this is a musig2 channel output, the keyspend is the only path we can take.
+func taprootKeyspendSighash(tx *wire.MsgTx, pkScript []byte,
+	value int64) ([]byte, error) {
+
+	prevOutputFetcher := txscript.NewCannedPrevOutputFetcher(
+		pkScript, value,
+	)
+
+	sigHashes := txscript.NewTxSigHashes(tx, prevOutputFetcher)
+
+	return txscript.CalcTaprootSignatureHash(
+		sigHashes, txscript.SigHashDefault, tx, 0, prevOutputFetcher,
+	)
+}
+
+// SignCommit signs the passed commitment w/ the current signing (relative
+// remote) nonce. Given nonces should only ever be used once, once the method
+// returns a new nonce is returned, w/ the existing nonce blanked out.
+func (m *MusigSession) SignCommit(tx *wire.MsgTx) (*MusigPartialSig, error) {
+	// If we already have a session, then we don't need to finalize as this
+	// was done up front (symmetric nonce case, like for co-op close).
+	if m.session == nil {
+		// Before we can sign a new commitment, we'll need to generate
+		// a fresh nonce that'll be sent along side our signature. With
+		// the nonce in hand, we can finalize the session.
+		//
+		// TODO(roasbeef): can also pass in stuff like the sighash to
+		// further bind context, etc, etc.
+		signingNonce, err := musig2.GenNonces(
+			musig2.WithPublicKey(m.localKey.PubKey),
+		)
+		if err != nil {
+			return nil, err
+		}
+		if err := m.FinalizeSession(*signingNonce); err != nil {
+			return nil, err
+		}
+	}
+
+	// Once we sign with a nonce, we'll never use it again, so it's safe to
+	// go ahead and clean up the session right here.
+	// defer m.signer.MuSig2Cleanup(m.session.SessionID)
+	//
+	// TODO(roasbeef): can't clean up here as need to combine sig
+
+	// Next we can sign, we'll need to generate the sighash for their
+	// commitment transaction.
+	sigHash, err := taprootKeyspendSighash(
+		tx, m.inputTxOut.PkScript, m.inputTxOut.Value,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// Now that we have our session created, we'll use it to generate the
+	// initial partial signature over our sighash.
+	var sigHashMsg [32]byte
+	copy(sigHashMsg[:], sigHash)
+
+	walletLog.Infof("Generating new musig2 sig for session=%x, nonces=%s",
+		m.session.SessionID[:], m.nonces.String())
+
+	sig, err := m.signer.MuSig2Sign(
+		m.session.SessionID, sigHashMsg, false,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewMusigPartialSig(
+		sig, m.session.PublicNonce, m.combinedNonce, m.signerKeys,
+	), nil
+}
+
+// Refresh is called once we receive a new verification nonce from the remote
+// party after sending a signature. This nonce will be coupled within the
+// revoke-and-ack message of the remote party.
```

Comment:
and coupled with `funding_locked` right?

Reply:
not sure I follow. can you maybe re-phrase? I thought we refresh the each time we get a verification nonce from the peer which would be if they send `funding_locked` or `channel_reestablish`

Reply:
For funding locked, later in the diff, a new set of functional options are used to init a channel with set musig2 nonce state. The `Refresh` method isn't called in that case, instead you're initializing the state for the first time, as it goes away after the TCP connection does. Only when verifying signatures and sending a revoke message is this called.

Reply:
cool - gotcha. Thanks :)